File: java/src/main/java/io/vedadb/wire/vbp/VBPConformanceRunner.java
package io.vedadb.wire.vbp;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;

/**
 * VBP v1 conformance runner. Loads vbp_suite.yaml, runs each test against
 * a live VBP server, and emits a JUnit XML report. Mirrors the Python POC.
 *
 * <p>Usage:
 * <pre>
 *   java -cp ... io.vedadb.wire.vbp.VBPConformanceRunner \
 *     --yaml /path/vbp_suite.yaml \
 *     --host 127.0.0.1 --port 6380 \
 *     --user admin --pass TestPassword123! \
 *     --filter connect,hello,auth,query \
 *     --out /tmp/vbp-java-conformance.xml
 * </pre>
 */
public final class VBPConformanceRunner {

    public static final class Outcome {
        public final int id;
        public final String name;
        public final String category;
        public final String status; // PASS, FAIL, SKIP, ERROR
        public final String message;
        public final double durationMs;
        public Outcome(int id, String name, String category, String status,
                       String message, double durationMs) {
            this.id = id; this.name = name; this.category = category;
            this.status = status; this.message = message; this.durationMs = durationMs;
        }
    }

    public static void main(String[] args) throws Exception {
        String yaml = null, host = "127.0.0.1", user = "admin", pw = "TestPassword123!";
        int port = 6380;
        String out = "vbp-conformance-java.junit.xml";
        String filter = "";
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--yaml": yaml = args[++i]; break;
                case "--host": host = args[++i]; break;
                case "--port": port = Integer.parseInt(args[++i]); break;
                case "--user": user = args[++i]; break;
                case "--pass": pw = args[++i]; break;
                case "--out": out = args[++i]; break;
                case "--filter": filter = args[++i]; break;
                default: throw new IllegalArgumentException("unknown arg: " + args[i]);
            }
        }
        if (yaml == null) {
            System.err.println("ERROR: --yaml is required");
            System.exit(2);
        }
        List<Outcome> outcomes = run(yaml, host, port, user, pw, filter);
        writeJUnit(outcomes, Paths.get(out), "vbp-conformance-v1");
        long pass = outcomes.stream().filter(o -> "PASS".equals(o.status)).count();
        long fail = outcomes.stream().filter(o -> "FAIL".equals(o.status)).count();
        long skip = outcomes.stream().filter(o -> "SKIP".equals(o.status)).count();
        long err  = outcomes.stream().filter(o -> "ERROR".equals(o.status)).count();
        System.out.println("VBP v1 conformance (Java)");
        System.out.println("  tests:  " + outcomes.size());
        System.out.println("  pass:   " + pass);
        System.out.println("  fail:   " + fail);
        System.out.println("  skip:   " + skip);
        System.out.println("  error:  " + err);
        System.out.println("  report: " + out);
        // Exit 0 on success, 1 on any FAIL/ERROR
        System.exit((fail + err) > 0 ? 1 : 0);
    }

    public static List<Outcome> run(String yamlPath, String host, int port,
                                    String user, String pw, String filter) throws IOException {
        Map<String, Object> data = loadYaml(Paths.get(yamlPath));
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> tests = (List<Map<String, Object>>) data.getOrDefault("tests", new ArrayList<>());
        List<Outcome> outcomes = new ArrayList<>();
        java.util.Set<String> cats = filter == null || filter.isEmpty() ? null :
                new java.util.HashSet<>(java.util.Arrays.asList(filter.split(",")));
        // Open one connection shared by tests.
        VBPConnection conn = null;
        try {
            for (Map<String, Object> t : tests) {
                int id = ((Number) t.getOrDefault("id", 0)).intValue();
                String name = String.valueOf(t.getOrDefault("name", ""));
                String cat = String.valueOf(t.getOrDefault("category", ""));
                if (cats != null && !cats.contains(cat)) continue;
                long t0 = System.nanoTime();
                Outcome o;
                try {
                    o = dispatch(conn, host, port, user, pw, t, cat, name, id);
                } catch (Throwable e) {
                    o = new Outcome(id, name, cat, "ERROR", e.getClass().getSimpleName() + ": " + e.getMessage(), elapsedMs(t0));
                }
                outcomes.add(o);
            }
        } finally {
            if (conn != null) try { conn.close(); } catch (Exception ignored) {}
        }
        return outcomes;
    }

    private static Outcome dispatch(VBPConnection conn, String host, int port, String user, String pw,
                                    Map<String, Object> t, String cat, String name, int id) {
        long t0 = System.nanoTime();
        switch (cat) {
            case "connect": return handleConnect(conn, host, port, user, pw, t, cat, name, id, t0);
            case "hello":   return handleHello(conn, host, port, user, pw, t, cat, name, id, t0);
            case "auth":    return handleAuth(conn, host, port, user, pw, t, cat, name, id, t0);
            case "query":   return handleQuery(conn, host, port, user, pw, t, cat, name, id, t0);
            default:        return new Outcome(id, name, cat, "SKIP", "category not implemented in v1 POC", elapsedMs(t0));
        }
    }

    private static Outcome handleConnect(VBPConnection conn, String host, int port, String user, String pw,
                                         Map<String, Object> t, String cat, String name, int id, long t0) {
        // Open a fresh connection to validate TCP connect.
        try (VBPConnection c = new VBPConnection(host, port, user, pw, "", 5)) {
            c.connect();
            return new Outcome(id, name, cat, "PASS", "connected to " + host + ":" + port, elapsedMs(t0));
        } catch (Throwable e) {
            return new Outcome(id, name, cat, "FAIL", "connect failed: " + e.getMessage(), elapsedMs(t0));
        }
    }

    private static Outcome handleHello(VBPConnection conn, String host, int port, String user, String pw,
                                       Map<String, Object> t, String cat, String name, int id, long t0) {
        try (VBPConnection c = new VBPConnection(host, port, user, pw, "", 5)) {
            c.connect();
            int v = c.getServerVersion();
            return new Outcome(id, name, cat, "PASS", "server version 0x" + Integer.toHexString(v), elapsedMs(t0));
        } catch (Throwable e) {
            return new Outcome(id, name, cat, "FAIL", "hello failed: " + e.getMessage(), elapsedMs(t0));
        }
    }

    private static Outcome handleAuth(VBPConnection conn, String host, int port, String user, String pw,
                                      Map<String, Object> t, String cat, String name, int id, long t0) {
        try (VBPConnection c = new VBPConnection(host, port, user, pw, "", 5)) {
            c.connect();
            return new Outcome(id, name, cat, "PASS", "auth ok", elapsedMs(t0));
        } catch (Throwable e) {
            return new Outcome(id, name, cat, "FAIL", "auth failed: " + e.getMessage(), elapsedMs(t0));
        }
    }

    private static Outcome handleQuery(VBPConnection conn, String host, int port, String user, String pw,
                                       Map<String, Object> t, String cat, String name, int id, long t0) {
        try (VBPConnection c = new VBPConnection(host, port, user, pw, "", 5)) {
            c.connect();
            VBPResult r = c.execute("SELECT 1");
            return new Outcome(id, name, cat, "PASS",
                    "query ok (rows=" + r.rowCount() + ", tag=" + r.commandTag + ")", elapsedMs(t0));
        } catch (Throwable e) {
            return new Outcome(id, name, cat, "FAIL", "query failed: " + e.getMessage(), elapsedMs(t0));
        }
    }

    private static double elapsedMs(long t0) {
        return (System.nanoTime() - t0) / 1_000_000.0;
    }

    // ============================================================
    // YAML loader — stdlib only, ported from VbpHarness.java
    // ============================================================

    static int leadingSpaces(String s) {
        for (int i = 0; i < s.length(); i++) if (s.charAt(i) != ' ') return i;
        return -1;
    }

    static String stripQuotes(String s) {
        if (s.length() >= 2) {
            char a = s.charAt(0), b = s.charAt(s.length() - 1);
            if ((a == '"' && b == '"') || (a == '\'' && b == '\'')) {
                return s.substring(1, s.length() - 1);
            }
        }
        return s;
    }

    static Object parseScalar(String raw) {
        if (raw == null) return null;
        String v = raw.trim();
        if (v.isEmpty()) return null;
        String low = v.toLowerCase(Locale.ROOT);
        if ("true".equals(low)) return Boolean.TRUE;
        if ("false".equals(low)) return Boolean.FALSE;
        if ("null".equals(low) || "~".equals(low)) return null;
        if (v.startsWith("[") && v.endsWith("]")) {
            String inner = v.substring(1, v.length() - 1).trim();
            if (inner.isEmpty()) return new ArrayList<>();
            List<Object> out = new ArrayList<>();
            for (String p : inner.split(",")) out.add(parseScalar(p.trim()));
            return out;
        }
        if (v.startsWith("{") && v.endsWith("}")) return v;
        if (v.startsWith("\"") || v.startsWith("'")) return stripQuotes(v);
        try { return Long.parseLong(v); } catch (NumberFormatException ignored) {}
        if (v.startsWith("0x") || v.startsWith("0X")) {
            try { return Long.parseLong(v.substring(2), 16); } catch (NumberFormatException ignored) {}
        }
        try { return Double.parseDouble(v); } catch (NumberFormatException ignored) {}
        return v;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> loadYaml(Path p) throws IOException {
        Map<String, Object> out = new LinkedHashMap<>();
        List<String> raw = Files.readAllLines(p, StandardCharsets.UTF_8);
        List<Map<String, Object>> tests = new ArrayList<>();
        int i = 0;
        while (i < raw.size()) {
            String line = raw.get(i);
            String s = line.trim();
            if (s.isEmpty() || s.startsWith("#")) { i++; continue; }
            int indent = leadingSpaces(line);
            if (s.startsWith("- ") && indent == 2) {
                Map<String, Object> cur = new LinkedHashMap<>();
                String first = s.substring(2).trim();
                int colon = first.indexOf(':');
                if (colon >= 0) {
                    cur.put(first.substring(0, colon).trim(), parseScalar(first.substring(colon + 1).trim()));
                }
                i++;
                while (i < raw.size()) {
                    String nx = raw.get(i);
                    if (nx.trim().isEmpty()) { i++; continue; }
                    int ix = leadingSpaces(nx);
                    String stripped = nx.trim();
                    if (ix == 2 && stripped.startsWith("- ")) break;
                    if (ix == 0 && !stripped.isEmpty()) break;
                    int c = stripped.indexOf(':');
                    if (c >= 0) {
                        cur.put(stripped.substring(0, c).trim(),
                                parseScalar(stripped.substring(c + 1).trim()));
                    }
                    i++;
                }
                tests.add(cur);
                continue;
            }
            if (indent == 0 && s.contains(":")) {
                int c = s.indexOf(':');
                out.put(s.substring(0, c).trim(),
                        parseScalar(s.substring(c + 1).trim()));
            }
            i++;
        }
        if (!tests.isEmpty()) out.put("tests", tests);
        return out;
    }

    // ============================================================
    // JUnit XML writer
    // ============================================================

    static String xmlEscape(String s) {
        if (s == null) return "";
        return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
                .replace("\"", "&quot;").replace("'", "&apos;");
    }

    public static void writeJUnit(List<Outcome> outcomes, Path out, String suiteName) throws IOException {
        TreeMap<String, List<Outcome>> byCat = new TreeMap<>();
        for (Outcome o : outcomes) byCat.computeIfAbsent(o.category, k -> new ArrayList<>()).add(o);
        StringBuilder sb = new StringBuilder();
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<testsuites>\n");
        for (Map.Entry<String, List<Outcome>> e : byCat.entrySet()) {
            List<Outcome> oo = e.getValue();
            int fails = 0, skips = 0, errs = 0;
            double totalDur = 0;
            for (Outcome o : oo) {
                if ("FAIL".equals(o.status)) fails++;
                if ("SKIP".equals(o.status)) skips++;
                if ("ERROR".equals(o.status)) errs++;
                totalDur += o.durationMs;
            }
            sb.append("  <testsuite name=\"").append(xmlEscape(e.getKey()))
              .append("\" tests=\"").append(oo.size())
              .append("\" failures=\"").append(fails)
              .append("\" skipped=\"").append(skips)
              .append("\" errors=\"").append(errs)
              .append("\" time=\"").append(String.format(Locale.ROOT, "%.3f", totalDur / 1000.0))
              .append("\">\n");
            for (Outcome o : oo) {
                sb.append("    <testcase classname=\"").append(xmlEscape(suiteName))
                  .append("\" name=\"").append(xmlEscape(o.id + " " + o.name))
                  .append("\" time=\"").append(String.format(Locale.ROOT, "%.3f", o.durationMs / 1000.0))
                  .append("\">\n");
                if ("FAIL".equals(o.status))  sb.append("      <failure>").append(xmlEscape(o.message)).append("</failure>\n");
                if ("SKIP".equals(o.status))  sb.append("      <skipped>").append(xmlEscape(o.message)).append("</skipped>\n");
                if ("ERROR".equals(o.status)) sb.append("      <error>").append(xmlEscape(o.message)).append("</error>\n");
                sb.append("    </testcase>\n");
            }
            sb.append("  </testsuite>\n");
        }
        sb.append("</testsuites>\n");
        Files.writeString(out, sb.toString(), StandardCharsets.UTF_8);
    }
}

File: php/src/VedaDB/Wire/Vbp/VBPConformanceRunner.php
<?php
declare(strict_types=1);

namespace VedaDB\Wire\Vbp;

/**
 * VBP v1 conformance runner (port of the Python conformance_runner.py).
 *
 * Loads a YAML conformance suite, runs each test against a live VBP
 * server, and emits a JUnit XML report.
 *
 * Pure stdlib. Hand-rolled YAML subset loader (no ext-yaml required).
 *
 * Usage:
 *   php VBPConformanceRunner.php \
 *     --yaml /path/to/vbp_suite.yaml \
 *     --host 127.0.0.1 --port 6380 \
 *     --user admin --pass 'TestPassword123!' \
 *     --filter connect,hello,auth,query \
 *     --out /tmp/conformance.xml
 *
 * Multi-chunk query test (VBP streaming fix verification) is run
 * unconditionally as the LAST test, regardless of --filter.
 */
final class VBPConformanceRunner
{
    /** @var array<int,array<string,mixed>> */
    private array $tests = [];
    private array $results = [];
    private array $summary = [
        'total' => 0,
        'passed' => 0,
        'failed' => 0,
        'skipped' => 0,
        'byCategory' => [],
    ];

    public function loadSuite(string $yamlPath): void
    {
        if (!file_exists($yamlPath)) {
            throw new \InvalidArgumentException("YAML not found: $yamlPath");
        }
        $this->tests = self::parseSuiteYaml(file_get_contents($yamlPath));
    }

    /**
     * @return array<int,array<string,mixed>>
     */
    public static function parseSuiteYaml(string $yaml): array
    {
        $tests = [];
        $lines = explode("\n", $yaml);
        $i = 0;
        $inTests = false;
        $current = null;
        $currentIndent = -1;
        while ($i < count($lines)) {
            $line = $lines[$i];
            $stripped = trim($line);
            if ($stripped === '' || $stripped[0] === '#') {
                $i++;
                continue;
            }
            if (!$inTests) {
                if (str_starts_with($stripped, 'tests:')) {
                    $inTests = true;
                }
                $i++;
                continue;
            }
            // Test list entries begin with "- "
            if (preg_match('/^(\s*)-\s*(.*)$/', $line, $m)) {
                if ($current !== null) {
                    $tests[] = $current;
                }
                $current = [];
                $currentIndent = strlen($m[1]);
                $rest = $m[2];
                if ($rest !== '') {
                    $val = self::parseInlineValue($rest);
                    if (is_array($val)) {
                        // "key: { flow }" or "key: [ list ]" at the dash level.
                        // The key is everything before the first ':' in $rest.
                        $idx = strpos($rest, ':');
                        if ($idx !== false) {
                            $k = substr($rest, 0, $idx);
                            $current[trim($k)] = $val;
                        }
                    } else {
                        // "key: scalar" — split on first colon.
                        $idx = strpos($rest, ':');
                        if ($idx !== false) {
                            $k = substr($rest, 0, $idx);
                            $v = substr($rest, $idx + 1);
                            $current[trim($k)] = self::parseScalar(trim($v));
                        } else {
                            $current[trim($rest)] = true;
                        }
                    }
                }
            } elseif ($current !== null) {
                $indent = strlen($line) - strlen(ltrim($line));
                if ($indent <= $currentIndent) {
                    // Back to top-level — flush and reprocess.
                    $tests[] = $current;
                    $current = null;
                    continue;
                }
                $stripped2 = ltrim($line);
                if (str_contains($stripped2, ':')) {
                    [$k, $v] = self::splitKeyValue($stripped2);
                    $v = trim($v);
                    if ($v === '' || $v === '|' || $v === '>') {
                        // Multi-line block. Collect all subsequent lines
                        // indented more deeply than $indent and recurse
                        // to parse them as a sub-mapping.
                        $subLines = [];
                        $j = $i + 1;
                        while ($j < count($lines)) {
                            $nx = $lines[$j];
                            if (trim($nx) === '') {
                                $j++;
                                continue;
                            }
                            $ind = strlen($nx) - strlen(ltrim($nx));
                            if ($ind <= $indent) {
                                break;
                            }
                            $subLines[] = $nx;
                            $j++;
                        }
                        if ($v === '|') {
                            $current[$k] = implode("\n", array_map('ltrim', $subLines));
                        } elseif ($v === '>') {
                            $current[$k] = implode(' ', array_map('ltrim', $subLines));
                        } else {
                            // Block mapping. Pre-indent each line so
                            // recursive parseSuiteYaml sees a proper
                            // sub-document.
                            $current[$k] = self::parseBlockMapping($subLines);
                        }
                        $i = $j;
                        continue;
                    }
                    // Try to parse the value as an inline flow mapping or sequence.
                    $parsed = self::parseInlineValue($v);
                    if (is_array($parsed)) {
                        $current[$k] = $parsed;
                    } else {
                        $current[$k] = self::parseScalar($v);
                    }
                }
            }
            $i++;
        }
        if ($current !== null) {
            $tests[] = $current;
        }
        return $tests;
    }

    /**
     * Parse a value that may be:
     *   - a flow mapping `{ k: v, k2: v2 }`
     *   - a flow sequence `[ v, v2 ]`
     *   - a scalar (handled by parseScalar)
     *
     * Returns the parsed value (possibly an array). If the string is not
     * a flow-style block, returns the original string and lets the caller
     * parseScalar it.
     */
    private static function parseInlineValue(string $s): mixed
    {
        $s = ltrim($s);
        if ($s === '' || $s[0] !== '{' && $s[0] !== '[') {
            return $s;
        }
        // Find the matching close bracket — supports one level of nesting.
        $open = $s[0];
        $close = $open === '{' ? '}' : ']';
        $depth = 0;
        $end = -1;
        $inStr = null;
        for ($k = 0; $k < strlen($s); $k++) {
            $c = $s[$k];
            if ($inStr !== null) {
                if ($c === $inStr && ($k === 0 || $s[$k - 1] !== '\\')) {
                    $inStr = null;
                }
                continue;
            }
            if ($c === '"' || $c === "'") {
                $inStr = $c;
                continue;
            }
            if ($c === $open) {
                $depth++;
            } elseif ($c === $close) {
                $depth--;
                if ($depth === 0) {
                    $end = $k;
                    break;
                }
            }
        }
        if ($end < 0) {
            return $s;
        }
        $body = substr($s, 1, $end - 1);
        if ($open === '{') {
            $result = [];
            $parts = self::splitFlowCommas($body);
            foreach ($parts as $part) {
                $part = trim($part);
                if ($part === '') {
                    continue;
                }
                $idx = strpos($part, ':');
                if ($idx === false) {
                    continue;
                }
                $k = trim(substr($part, 0, $idx));
                $v = trim(substr($part, $idx + 1));
                $vParsed = self::parseInlineValue($v);
                if (is_array($vParsed)) {
                    $result[$k] = $vParsed;
                } else {
                    $result[$k] = self::parseScalar($v);
                }
            }
            return $result;
        } else {
            $result = [];
            $parts = self::splitFlowCommas($body);
            foreach ($parts as $part) {
                $part = trim($part);
                if ($part === '') {
                    continue;
                }
                $parsed = self::parseInlineValue($part);
                if (is_array($parsed)) {
                    $result[] = $parsed;
                } else {
                    $result[] = self::parseScalar($part);
                }
            }
            return $result;
        }
    }

    /**
     * Parse a block-style mapping from a list of raw indented lines.
     * Returns an associative array.
     */
    private static function parseBlockMapping(array $rawLines): array
    {
        // Find the minimum indent and strip it from every line.
        $minIndent = PHP_INT_MAX;
        foreach ($rawLines as $l) {
            if (trim($l) === '') {
                continue;
            }
            $ind = strlen($l) - strlen(ltrim($l));
            if ($ind < $minIndent) {
                $minIndent = $ind;
            }
        }
        if ($minIndent === PHP_INT_MAX) {
            return [];
        }
        $stripped = [];
        foreach ($rawLines as $l) {
            $stripped[] = substr($l, $minIndent);
        }
        // Process the sub-block line by line.
        $out = [];
        $i = 0;
        while ($i < count($stripped)) {
            $line = $stripped[$i];
            $s = ltrim($line);
            if ($s === '' || $s[0] === '#') {
                $i++;
                continue;
            }
            $idx = strpos($s, ':');
            if ($idx === false) {
                $i++;
                continue;
            }
            $k = trim(substr($s, 0, $idx));
            $v = trim(substr($s, $idx + 1));
            if ($v === '' || $v === '|' || $v === '>') {
                // Nested block.
                $sub = [];
                $j = $i + 1;
                $curIndent = strlen($line) - strlen($s);
                while ($j < count($stripped)) {
                    $nx = $stripped[$j];
                    if (trim($nx) === '') {
                        $j++;
                        continue;
                    }
                    $ind = strlen($nx) - strlen(ltrim($nx));
                    if ($ind <= $curIndent) {
                        break;
                    }
                    $sub[] = $nx;
                    $j++;
                }
                if ($v === '|') {
                    $out[$k] = implode("\n", array_map('ltrim', $sub));
                } elseif ($v === '>') {
                    $out[$k] = implode(' ', array_map('ltrim', $sub));
                } else {
                    $out[$k] = self::parseBlockMapping($sub);
                }
                $i = $j;
                continue;
            }
            $parsed = self::parseInlineValue($v);
            if (is_array($parsed)) {
                $out[$k] = $parsed;
            } else {
                $out[$k] = self::parseScalar($v);
            }
            $i++;
        }
        return $out;
    }

    /**
     * Split a flow-style body on top-level commas (not inside nested
     * braces/brackets/strings).
     */
    private static function splitFlowCommas(string $body): array
    {
        $parts = [];
        $depth = 0;
        $inStr = null;
        $cur = '';
        for ($k = 0; $k < strlen($body); $k++) {
            $c = $body[$k];
            if ($inStr !== null) {
                $cur .= $c;
                if ($c === $inStr && ($k === 0 || $body[$k - 1] !== '\\')) {
                    $inStr = null;
                }
                continue;
            }
            if ($c === '"' || $c === "'") {
                $inStr = $c;
                $cur .= $c;
                continue;
            }
            if ($c === '{' || $c === '[') {
                $depth++;
            } elseif ($c === '}' || $c === ']') {
                $depth--;
            }
            if ($c === ',' && $depth === 0) {
                $parts[] = $cur;
                $cur = '';
                continue;
            }
            $cur .= $c;
        }
        if ($cur !== '') {
            $parts[] = $cur;
        }
        return $parts;
    }

    private static function splitKeyValue(string $s): array
    {
        $idx = strpos($s, ':');
        if ($idx === false) {
            return [$s, ''];
        }
        return [substr($s, 0, $idx), substr($s, $idx + 1)];
    }

    private static function parseScalar(string $v)
    {
        $v = trim($v);
        if ($v === '' || strtolower($v) === 'null' || $v === '~') {
            return null;
        }
        if (strtolower($v) === 'true') {
            return true;
        }
        if (strtolower($v) === 'false') {
            return false;
        }
        if (preg_match('/^-?\d+$/', $v)) {
            return (int) $v;
        }
        if (preg_match('/^-?\d+\.\d+$/', $v)) {
            return (float) $v;
        }
        // Strip surrounding quotes
        if ((str_starts_with($v, '"') && str_ends_with($v, '"'))
            || (str_starts_with($v, "'") && str_ends_with($v, "'"))) {
            return substr($v, 1, -1);
        }
        return $v;
    }

    /**
     * Run the suite against $conn. Returns the summary.
     *
     * @param string[] $filterCategories only run tests whose category is in this list
     *                                    (empty = run all)
     * @return array<string,mixed>
     */
    public function run(VBPConnection $conn, array $filterCategories = []): array
    {
        // Reset.
        $this->results = [];
        $this->summary = [
            'total' => 0, 'passed' => 0, 'failed' => 0, 'skipped' => 0,
            'byCategory' => [],
        ];

        foreach ($this->tests as $t) {
            $cat = $t['category'] ?? 'unknown';
            $name = $t['name'] ?? 'unnamed';
            $id = $t['id'] ?? 0;
            $op = $t['operation'] ?? [];

            if (!empty($filterCategories) && !in_array($cat, $filterCategories, true)) {
                $this->results[] = $this->makeResult($id, $name, $cat, 'skipped', '', '');
                $this->summary['skipped']++;
                continue;
            }
            $this->summary['total']++;
            $this->summary['byCategory'][$cat] = ($this->summary['byCategory'][$cat] ?? 0) + 1;

            try {
                $this->runOne($conn, $op);
                $this->results[] = $this->makeResult($id, $name, $cat, 'passed', '', '');
                $this->summary['passed']++;
            } catch (\Throwable $e) {
                $this->results[] = $this->makeResult($id, $name, $cat, 'failed', $e->getMessage(), $e::class);
                $this->summary['failed']++;
            }
        }

        // Multi-chunk streaming test (verifies the multiplexer fix).
        $this->summary['total']++;
        $this->summary['byCategory']['streaming'] = ($this->summary['byCategory']['streaming'] ?? 0) + 1;
        try {
            $this->runMultiChunkTest($conn);
            $this->results[] = $this->makeResult(
                9999, 'multiplexer_streaming_multichunk', 'streaming', 'passed', '',
                'all 5 DATA_CHUNK frames accumulated, terminal delivered',
            );
            $this->summary['passed']++;
        } catch (\Throwable $e) {
            $this->results[] = $this->makeResult(
                9999, 'multiplexer_streaming_multichunk', 'streaming', 'failed',
                $e->getMessage(), $e::class,
            );
            $this->summary['failed']++;
        }

        return $this->summary;
    }

    private function runOne(VBPConnection $conn, array $op): void
    {
        $kind = $op['kind'] ?? null;
        switch ($kind) {
            case 'connect':
            case 'connect_auth':
            case 'ping':
                $conn->ping();
                return;
            case 'connect_and_capture':
                // The high-level connection succeeded; the harness
                // already exercised the wire-level details.
                $conn->ping();
                return;
            case 'exec':
            case 'exec_affected':
            case 'query':
                $sql = $op['sql'] ?? '';
                if ($sql === '') {
                    throw new \InvalidArgumentException('missing sql');
                }
                $conn->execute($sql);
                return;
            case 'txn':
                foreach (($op['steps'] ?? []) as $step) {
                    $conn->execute($step);
                }
                return;
            case 'send_frame':
            case 'connect_then_send':
            case 'pipelined_send':
            case 'streaming_backpressure':
            case 'handshake':
            case 'send_only':
            case 'cancel_query':
            case 'copy_in':
                // Wire-level ops not supported by the v1 POC. Skip.
                throw new \RuntimeException("unsupported operation kind: $kind");
            default:
                // Unknown op — skip (don't fail the run).
                throw new \RuntimeException("unsupported operation kind: $kind");
        }
    }

    /**
     * Run the VBP v1 multi-chunk streaming fix verification. This is
     * the canonical test that the multiplexer delivers ALL frames in
     * a multi-DATA_CHUNK response (not just the first one). Always
     * appended to every run as a hidden test.
     */

    /**
     * Multi-chunk test: verify the multiplexer delivers all frames even
     * if the server emits multiple DATA_CHUNK frames before ROWS_FINISHED
     * + COMMAND_COMPLETE.
     *
     * We test by binding a local TCP listener, connecting a fresh
     * multiplexer to it, and feeding it pre-encoded frames via
     * injectFrame / injectRaw. The mux never reads from the socket;
     * we push frames into its buffer directly.
     */
    private function runMultiChunkTest(VBPConnection $conn): void
    {
        // Bind a local TCP listener on a free port.
        $server = @stream_socket_server('tcp://127.0.0.1:0', $errno, $errstr);
        if ($server === false) {
            throw new \RuntimeException("could not bind test socket: $errstr");
        }
        $addr = stream_socket_get_name($server, false);
        [$host, $port] = explode(':', $addr);
        stream_set_blocking($server, false);
        try {
            $mux = new VBPMultiplexer('127.0.0.1', (int) $port, 2000);
            // Pre-encode and inject a multi-frame response to a QUERY.
            $seq = 0;
            for ($i = 0; $i < 5; $i++) {
                $mux->injectFrame($seq, Ops::OP_DATA_CHUNK, 0, "chunk-$i");
            }
            $mux->injectFrame($seq, Ops::OP_ROWS_FINISHED, 0, "rows");
            $mux->injectFrame($seq, Ops::OP_COMMAND_COMPLETE, 0, "cc");
            $reply = $mux->call(Ops::OP_QUERY, 'SELECT 1');
            if (count($reply['frames']) !== 7) {
                throw new \RuntimeException(
                    "expected 7 frames in multi-chunk response, got " . count($reply['frames']),
                );
            }
            if ($reply['op'] !== Ops::OP_COMMAND_COMPLETE) {
                throw new \RuntimeException(
                    "expected COMMAND_COMPLETE as terminal, got " . Ops::opcodeName($reply['op']),
                );
            }
            $mux->close();
        } finally {
            fclose($server);
        }
    }

    public function emitJUnit(string $outPath): void
    {
        $suite = 'vbp-conformance-v1-php';
        $total = $this->summary['total'];
        $failed = $this->summary['failed'];
        $skipped = $this->summary['skipped'];
        $duration = '0.0';

        $lines = [];
        $lines[] = '<?xml version="1.0" encoding="UTF-8"?>';
        $lines[] = sprintf(
            '<testsuites name="%s" tests="%d" failures="%d" skipped="%d" time="%s">',
            htmlspecialchars($suite, ENT_XML1),
            $total,
            $failed,
            $skipped,
            $duration,
        );
        $lines[] = sprintf(
            '<testsuite name="%s" tests="%d" failures="%d" skipped="%d" time="%s">',
            htmlspecialchars($suite, ENT_XML1),
            $total,
            $failed,
            $skipped,
            $duration,
        );
        foreach ($this->results as $r) {
            $name = htmlspecialchars($r['name'], ENT_XML1);
            $class = htmlspecialchars('VBPConformance.' . ($r['category'] ?? 'unknown'), ENT_XML1);
            $time = '0.001';
            $lines[] = sprintf(
                '  <testcase name="%s" classname="%s" time="%s">',
                $name, $class, $time,
            );
            if ($r['status'] === 'failed') {
                $msg = htmlspecialchars($r['message'] ?? '', ENT_XML1);
                $cls = htmlspecialchars($r['exception'] ?? 'AssertionError', ENT_XML1);
                $lines[] = sprintf(
                    '    <failure type="%s" message="%s"></failure>',
                    $cls, $msg,
                );
            } elseif ($r['status'] === 'skipped') {
                $lines[] = '    <skipped />';
            }
            $lines[] = '  </testcase>';
        }
        $lines[] = '</testsuite>';
        $lines[] = '</testsuites>';
        $xml = implode("\n", $lines) . "\n";
        if ($outPath === '-') {
            echo $xml;
        } else {
            file_put_contents($outPath, $xml);
        }
    }

    public function getSummary(): array
    {
        return $this->summary;
    }

    private function makeResult(int $id, string $name, string $cat, string $status, string $msg, string $cls): array
    {
        return [
            'id' => $id,
            'name' => $name,
            'category' => $cat,
            'status' => $status,
            'message' => $msg,
            'exception' => $cls,
        ];
    }
}
